### PR TITLE
Move function unique ID to abi-tags

### PR DESCRIPTION
**Pull request by sklam**
- Add unique-id to `FunctionIdentity` and `FunctionDescriptor` classes.
- `RecursionCall` class API is improved. 
- Mangler function now takes an optional `uid` argument for the unique id.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
@sklam This patch updates and has a new test for this change, feel free to incorporate etc: https://github.com/stuartarchibald/numba/commit/81fad4f5f859a162396beaee16f51c9d114f69aa

**Comment by sklam** [comment]
@stuartarchibald, i cherrypicked your tests. (It's showing up before your message about it in this page.)

File: numba/tests/gdb/test_break_on_symbol_version.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# NOTE: This test is sensitive to line numbers as it checks breakpoints
+from numba import njit
```
Comment (stuartarchibald) [inline numba/tests/gdb/test_break_on_symbol_version.py:2]: @sklam as I wrote this, please can you review the test file? Thanks.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch @sklam. Two comments 1. is inline, about reviewing the test I wrote. 2. Should an example be added to the docs?

File: numba/tests/gdb/test_break_on_symbol_version.py (hunk 1 above)
Comment (sklam) [reply numba/tests/gdb/test_break_on_symbol_version.py:2]: Yes, i have reviewed it and play around with it as i cherrypicked it over.

**Comment by sklam** [comment]
@stuartarchibald, regarding docs and example, I think it's better if we convert the gdb docs to a more generative approach once we merge #76er 17, #7764 and #7771 (this). I find the current docs quite scattered and requires too much manual effort each time we the debuginfo.

File: numba/tests/gdb/test_break_on_symbol_version.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/tests/gdb/test_break_on_symbol_version.py:2]: Great, thanks for confirming.

**Comment by stuartarchibald** [comment]
> @stuartarchibald, regarding docs and example, I think it's better if we convert the gdb docs to a more generative approach once we merge #7617, #7764 and #7771 (this). I find the current docs quite scattered and requires too much manual effort each time we the debuginfo.

Having manually updated this region of the docs a few times, I agree, a more generative approach would be beneficial. I also think it might be worth having a section on native debugging that is stand-alone. Let's leave doc changes for now with view of getting this functionality merged.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch!

**Pull request merged**